File: src/algorithms/trees/bst-operations/bst-insert/__tests__/bst-insert_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))
module = importlib.import_module("bst-insert")
BSTNode = module.BSTNode
bst_insert = module.bst_insert


def make_node(value, left=None, right=None):
    node = BSTNode(value)
    node.left = left
    node.right = right
    return node


tree = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))


def test_inserts_greater_than_all():
    fresh = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))
    result = bst_insert(fresh, 8)
    assert result.right.right.right.value == 8


def test_inserts_into_left_subtree():
    fresh = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))
    result = bst_insert(fresh, 0)
    assert result.left.left.left.value == 0


def test_creates_root_from_none():
    result = bst_insert(None, 10)
    assert result.value == 10


def test_ignores_duplicates():
    fresh = make_node(4, make_node(2), make_node(6))
    result = bst_insert(fresh, 4)
    assert result.value == 4


if __name__ == "__main__":
    test_inserts_greater_than_all()
    test_inserts_into_left_subtree()
    test_creates_root_from_none()
    test_ignores_duplicates()
    print("All tests passed!")

File: src/algorithms/trees/bst-operations/bst-insert/__tests__/bst-insert_test.rs
include!("../sources/bst-insert.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<BSTNode>>, right: Option<Box<BSTNode>>) -> Option<Box<BSTNode>> {
        Some(Box::new(BSTNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<BSTNode>> {
        make_node(value, None, None)
    }

    fn build_tree() -> Option<Box<BSTNode>> {
        make_node(4,
            make_node(2, leaf(1), leaf(3)),
            make_node(6, leaf(5), leaf(7)),
        )
    }

    #[test]
    fn test_inserts_greater_than_all() {
        let result = bst_insert(build_tree(), 8);
        assert_eq!(result.right.as_ref().unwrap().right.as_ref().unwrap().right.as_ref().unwrap().value, 8);
    }

    #[test]
    fn test_inserts_into_left_subtree() {
        let result = bst_insert(build_tree(), 0);
        assert_eq!(result.left.as_ref().unwrap().left.as_ref().unwrap().left.as_ref().unwrap().value, 0);
    }

    #[test]
    fn test_creates_root_from_none() {
        let result = bst_insert(None, 10);
        assert_eq!(result.value, 10);
    }

    #[test]
    fn test_ignores_duplicates() {
        let tree = make_node(4, leaf(2), leaf(6));
        let result = bst_insert(tree, 4);
        assert_eq!(result.value, 4);
    }
}
